nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/StandardFlowSynchronizer.java

```diff
@@ -196,322 +178,343 @@ public void sync(final FlowController controller, final DataFlow proposedFlow, f
+            final FlowInheritability inheritability = fingerprintCheck.checkInheritability(existingDataFlow, proposedFlow, controller);
+
+            if (inheritability.isInheritable()) {
+                logger.debug("Proposed flow is inheritable");
             } else {
-                existingFlow = readFlowFromDisk();
-                if (existingFlow == null || existingFlow.length == 0) {
-                    existingFlowEmpty = true;
+                if (controller.isInitialized()) {
+                    // Flow has already been initialized so cannot inherit the cluster flow as liberally.
+                    // Since the cluster's flow is not immediately inheritable, we must throw an UninheritableFlowException.
+                    throw new UninheritableFlowException("Proposed configuration is not inheritable by the flow controller because of flow differences: " + inheritability.getExplanation());
+                }
+
+                logger.debug("Proposed flow is not directly inheritable. However, the Controller has not been synchronized yet, " +
+                    "so will check if the existing flow can be backed up and replaced by the proposed flow.");
+
+                final FlowInheritabilityCheck connectionMissingCheck = new ConnectionMissingCheck();
+                final FlowInheritability connectionMissingInheritability = connectionMissingCheck.checkInheritability(existingDataFlow, proposedFlow, controller);
+                if (connectionMissingInheritability.isInheritable()) {
+                    backupAndPurge = true;
+                    existingFlowEmpty = true; // Consider the existing flow as being empty
+                    logger.debug("Proposed flow contains all connections that currently have data queued. Will backup existing flow and replace, provided all other checks pass");
                 } else {
-                    final Document document = parseFlowBytes(existingFlow);
-                    final Element rootElement = document.getDocumentElement();
-                    final FlowEncodingVersion encodingVersion = FlowEncodingVersion.parse(rootElement);
-
-                    logger.trace("Setting controller thread counts");
-                    final Integer maxThreadCount = getInteger(rootElement, "maxThreadCount");
-                    if (maxThreadCount == null) {
-                        controller.setMaxTimerDrivenThreadCount(getInt(rootElement, "maxTimerDrivenThreadCount"));
-                        controller.setMaxEventDrivenThreadCount(getInt(rootElement, "maxEventDrivenThreadCount"));
-                    } else {
-                        controller.setMaxTimerDrivenThreadCount(maxThreadCount * 2 / 3);
-                        controller.setMaxEventDrivenThreadCount(maxThreadCount / 3);
-                    }
+                    throw new UninheritableFlowException("Proposed flow is not inheritable by the flow controller and cannot completely replace the current flow due to: "
```

Comment:
Since we know this exception is being thrown because the local flow has flow files in queues that don't exist in the cluster, could we make this message say something like that?

The exception I got for this case is below and seems like it would be hard for a user to know what issue was, they would have to take one of those UUIDs and know how to figure out what it is and why the fingerprint was different...

```
2020-04-07 13:29:45,831 ERROR [main] o.a.nifi.controller.StandardFlowService Failed to load flow from cluster due to: org.apache.nifi.controller.UninheritableFlowException: Failed to connect node to cluster because local flow is different than cluster flow.
org.apache.nifi.controller.UninheritableFlowException: Failed to connect node to cluster because local flow is different than cluster flow.
	at org.apache.nifi.controller.StandardFlowService.loadFromConnectionResponse(StandardFlowService.java:1007)
	at org.apache.nifi.controller.StandardFlowService.load(StandardFlowService.java:519)
	at org.apache.nifi.web.server.JettyServer.start(JettyServer.java:1071)
	at org.apache.nifi.NiFi.<init>(NiFi.java:158)
	at org.apache.nifi.NiFi.<init>(NiFi.java:72)
	at org.apache.nifi.NiFi.main(NiFi.java:301)
Caused by: org.apache.nifi.controller.UninheritableFlowException: Proposed flow is not inheritable by the flow controller and cannot completely replace the current flow due to: Found difference in Flows:
Local Fingerprint contains additional configuration from Cluster Fingerprint: 555052fa-0171-1000-ffff-ffffdb79d348NO_VALUE55502324-0171-1000-ffff-ffffd1ceb4e4554c8b10-0171-1000-6f7b-c2e4677be50aPROCESSOR55503fff-0171-1000-ffff-ffffdad8f473554c8b10-0171-1000-6f7b-c2e4677be50aPRO
	at org.apache.nifi.controller.StandardFlowSynchronizer.sync(StandardFlowSynchronizer.java:218)
	at org.apache.nifi.controller.FlowController.synchronize(FlowController.java:1426)
	at org.apache.nifi.persistence.StandardXMLFlowConfigurationDAO.load(StandardXMLFlowConfigurationDAO.java:89)
	at org.apache.nifi.controller.StandardFlowService.loadFromBytes(StandardFlowService.java:792)
	at org.apache.nifi.controller.StandardFlowService.loadFromConnectionResponse(StandardFlowService.java:985)
	... 5 common frames omitted
```

Reply:
Actually from reviewing the code further, I think just changing from `inheritability.getExplanation()` to `connectionMissingInheritability.getExplanation()` will give us what I was suggesting

Reply:
Ah yeah, looks like that was just a typo. Good catch on the bug :)